File: native/managers/socket/event_emitter.py
"""
Event Emitter module for publish-subscribe pattern.
Provides SocketEvent enum and EventEmitter base class.
"""

from enum import Enum
from typing import Optional, Dict, Set, Callable, Any


class SocketEvent(Enum):
    """Socket event types"""
    CONNECT = "connect"
    DISCONNECT = "disconnect"
    MESSAGE = "message"
    ERROR = "error"
    ACTIONS_RECEIVED = "actions_received"


# Event listener type alias
EventListener = Callable[..., None]


class EventEmitter:
    """
    Event emitter implementing publish-subscribe pattern.
    Provides event subscription and publishing functionality.
    """

    def __init__(self):
        """Initialize event emitter"""
        self.events: Dict[SocketEvent, Set[EventListener]] = {}

    def on(self, event: SocketEvent, listener: EventListener):
        """
        Subscribe to an event.

        Args:
            event: Event type
            listener: Event listener function
        """
        if event not in self.events:
            self.events[event] = set()
        self.events[event].add(listener)

    def off(self, event: SocketEvent, listener: EventListener):
        """
        Unsubscribe from an event.

        Args:
            event: Event type
            listener: Event listener function
        """
        if event in self.events:
            self.events[event].discard(listener)

    def emit(self, event: SocketEvent, *args, **kwargs):
        """
        Publish an event.

        Args:
            event: Event type
            *args: Positional arguments
            **kwargs: Keyword arguments
        """
        if event in self.events:
            # Copy list to avoid modification during iteration
            for listener in list(self.events[event]):
                try:
                    listener(*args, **kwargs)
                except Exception as error:
                    print(f"Error in event listener for {event.value}: {error}")

    def remove_all_listeners(self, event: Optional[SocketEvent] = None):
        """
        Remove all listeners.

        Args:
            event: If specified, only remove listeners for this event;
                   otherwise remove all listeners
        """
        if event:
            self.events.pop(event, None)
        else:
            self.events.clear()

File: frontend/src/game/managers/socket/EventEmitter.ts
/**
 * EventEmitter - 事件发射器
 * 实现发布订阅模式
 */
import { SocketEvent } from '../SocketManager'

// 事件监听器类型
type EventListener = (...args: unknown[]) => void

/**
 * 事件发射器（发布订阅模式）
 */
export class EventEmitter {
  private events: Map<SocketEvent, Set<EventListener>> = new Map()

  /**
   * 订阅事件
   */
  on(event: SocketEvent, listener: EventListener): void {
    if (!this.events.has(event)) {
      this.events.set(event, new Set())
    }
    this.events.get(event)!.add(listener)
    console.log(`[EventEmitter] ✅ 订阅事件: ${event}, 当前监听器数: ${this.events.get(event)!.size}`)
  }

  /**
   * 取消订阅
   */
  off(event: SocketEvent, listener: EventListener): void {
    const listeners = this.events.get(event)
    if (listeners) {
      listeners.delete(listener)
    }
  }

  /**
   * 发布事件
   */
  emit(event: SocketEvent, ...args: unknown[]): void {
    const listeners = this.events.get(event)
    if (listeners && listeners.size > 0) {
      console.log(`[EventEmitter] 📤 发布事件: ${event}, 监听器数量: ${listeners.size}`)
      listeners.forEach(listener => {
        try {
          listener(...args)
        } catch (error) {
          console.error(`Error in event listener for ${event}:`, error)
        }
      })
    } else {
      console.warn(`[EventEmitter] ⚠️ 没有监听器订阅事件: ${event}`)
    }
  }

  /**
   * 清除所有监听器
   */
  removeAllListeners(event?: SocketEvent): void {
    if (event) {
      this.events.delete(event)
    } else {
      this.events.clear()
    }
  }
}
